Validation/MuonHits/src/ME0SimHitMatcher.cc

```diff
@@ -0,0 +1,215 @@
+#include "Validation/MuonHits/interface/ME0SimHitMatcher.h"
+
+using namespace std;
+
+ME0SimHitMatcher::ME0SimHitMatcher(const edm::ParameterSet& ps,
+                                   edm::ConsumesCollector&& iC)
+    : MuonSimHitMatcher(ps, std::move(iC)) {
+  simHitPSet_ = ps.getParameterSet("me0SimHit");
+  verbose_ = simHitPSet_.getParameter<int>("verbose");
+  simMuOnly_ = simHitPSet_.getParameter<bool>("simMuOnly");
+  discardEleHits_ = simHitPSet_.getParameter<bool>("discardEleHits");
+
+  simHitInput_ = iC.consumes<edm::PSimHitContainer>(
+      simHitPSet_.getParameter<edm::InputTag>("inputTag"));
+}
+
+/// initialize the event
+void ME0SimHitMatcher::init(const edm::Event& iEvent,
+                            const edm::EventSetup& iSetup) {
+  iSetup.get<MuonGeometryRecord>().get(me0_geom_);
+  if (me0_geom_.isValid()) {
+    geometry_ = &*me0_geom_;
+  } else {
+    hasGeometry_ = false;
+    edm::LogWarning("ME0SimHitMatcher")
+        << "+++ Info: ME0 geometry is unavailable. +++\n";
+  }
+  MuonSimHitMatcher::init(iEvent, iSetup);
+}
+
+/// do the matching
+void ME0SimHitMatcher::match(const SimTrack& track, const SimVertex& vertex) {
+  MuonSimHitMatcher::match(track, vertex);
+
+  if (hasGeometry_) {
+    matchSimHitsToSimTrack(track_ids_, simHits_);
+
+    if (verbose_) {
+      edm::LogInfo("ME0SimHitMatcher")
+          << "nTrackIds " << track_ids_.size() << " nSelectedME0SimHits "
+          << hits_.size() << endl;
+      edm::LogInfo("ME0SimHitMatcher")
+          << "detids ME0 " << detIds().size() << endl;
+
+      const auto& me0_ch_ids = detIds();
+      for (const auto& id : me0_ch_ids) {
+        const auto& me0_simhits = MuonSimHitMatcher::hitsInChamber(id);
+        const auto& me0_simhits_gp = simHitsMeanPosition(me0_simhits);
+        edm::LogInfo("ME0SimHitMatcher")
+            << "me0chid " << ME0DetId(id) << ": nHits " << me0_simhits.size()
+            << " phi " << me0_simhits_gp.phi() << " nCh "
+            << chamber_to_hits_[id].size() << endl;
+        const auto& strips = hitStripsInDetId(id);
+        edm::LogInfo("ME0SimHitMatcher") << "nStrip " << strips.size() << endl;
+        edm::LogInfo("ME0SimHitMatcher") << "strips : ";
+        std::copy(strips.begin(), strips.end(),
+                  ostream_iterator<int>(cout, " "));
```

Comment:
This will also write to `cout` which is forbidden. Could you please substitute this call with `MessageLogger` as well?